### ISSUE TITLE
grafana_datasource doesn't set password correctly

### ISSUE THREAD

**Issue by onefourfive** [ansible-collections/community.grafana#113]
##### SUMMARY

The `grafana_datasource` module does not configure the datasource password correctly. When used, the 'password' parameter is not used in the specified datasource. This might be due to the password being specified outside of the 'secureJsonData' parameter in the Grafana API request. See [Grafana API docs on Creating a Datasource.](https://grafana.com/docs/grafana/latest/http_api/data_source/#create-a-data-source)
##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

grafana_datasource
##### ANSIBLE VERSION

```paste below
ansible 2.9.9
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/derek/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/derek/.local/lib/python3.6/site-packages/ansible
  executable location = /home/derek/.local/bin/ansible
  python version = 3.6.9 (default, Jul 17 2020, 12:50:27) [GCC 8.4.0]
```

##### CONFIGURATION

```paste below
(no output)
```

##### OS / ENVIRONMENT

Ubuntu 18.04 LTS

##### STEPS TO REPRODUCE

Use `grafana_datasource` to create a MySQL datasource with MySQL username and password specified using the `user` and `password` parameters.

```yaml
  - name: provisioning datasources
    grafana_datasource:
      name: "{{ item.name }}"
      database: "{{ item.database}}"
      url: dsurl:port
      grafana_url: "{{ hostvars[inventory_hostname].grafurl }}"
      validate_certs: no
      grafana_api_key: "{{ hostvars[inventory_hostname].grafapikey }}"
      ds_type: mysql
      # any of the following reproduces the issue:
      user: testuser
      password: test123
      # 
      basic_auth_user: testuser
      basic_auth_password: test123
      # 
      basic_auth_user: testuser
      password: test123
      # 
```

##### EXPECTED RESULTS

The password is set and the datasource can be tested from the Grafana UI. The password field appears as 'configured' in a grayed out box. This matches behavior when the password is set through the UI.

##### ACTUAL RESULTS

The username is set however the password field remains blank in the UI and the datasource cannot be tested or used. The password must be set manually through the UI.

```paste below
ansible@host:~/$ ansible host1.net -m grafana_datasource -a "name=authtest url=https://authtest:3306 access=proxy state=present grafana_url=https://host1.net validate_certs=no grafana_api_key=loTsOfLettErS ds_type=mysql database=testdb basic_auth_password=test123 basic_auth_user=testuser" -K -vvv
ansible 2.9.9
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/ansible/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ansible/.local/lib/python3.6/site-packages/ansible
  executable location = /home/ansible/.local/bin/ansible
  python version = 3.6.9 (default, Jul 17 2020, 12:50:27) [GCC 8.4.0]
Using /etc/ansible/ansible.cfg as config file
BECOME password:
host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Parsed /etc/ansible/hosts inventory source with ini plugin
META: ran handlers
<host1.net> ESTABLISH SSH CONNECTION FOR USER: None
<host1.net> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/ansible/.ansible/cp/9ec4a2bc8e host1.net '/bin/sh -c '"'"'echo ~ && sleep 0'"'"''
<host1.net> (0, b'/home/ansible\n', b'')
<host1.net> ESTABLISH SSH CONNECTION FOR USER: None
<host1.net> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/ansible/.ansible/cp/9ec4a2bc8e host1.net '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /home/ansible/.ansible/tmp `"&& mkdir /home/ansible/.ansible/tmp/ansible-tmp-1596772708.4799106-4131-56466493012576 && echo ansible-tmp-1596772708.4799106-4131-56466493012576="` echo /home/ansible/.ansible/tmp/ansible-tmp-1596772708.4799106-4131-56466493012576 `" ) && sleep 0'"'"''
<host1.net> (0, b'ansible-tmp-1596772708.4799106-4131-56466493012576=/home/ansible/.ansible/tmp/ansible-tmp-1596772708.4799106-4131-56466493012576\n', b'')
Using module file /home/ansible/.local/lib/python3.6/site-packages/ansible/modules/monitoring/grafana_datasource.py
<host1.net> PUT /home/ansible/.ansible/tmp/ansible-local-4126tv0mzdhb/tmp50x_scxa TO /home/ansible/.ansible/tmp/ansible-tmp-1596772708.4799106-4131-56466493012576/AnsiballZ_grafana_datasource.py
<host1.net> SSH: EXEC sftp -b - -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/ansible/.ansible/cp/9ec4a2bc8e '[host1.net]'
<host1.net> (0, b'sftp> put /home/ansible/.ansible/tmp/ansible-local-4126tv0mzdhb/tmp50x_scxa /home/ansible/.ansible/tmp/ansible-tmp-1596772708.4799106-4131-56466493012576/AnsiballZ_grafana_datasource.py\n', b'')
<host1.net> ESTABLISH SSH CONNECTION FOR USER: None
<host1.net> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/ansible/.ansible/cp/9ec4a2bc8e host1.net '/bin/sh -c '"'"'chmod u+x /home/ansible/.ansible/tmp/ansible-tmp-1596772708.4799106-4131-56466493012576/ /home/ansible/.ansible/tmp/ansible-tmp-1596772708.4799106-4131-56466493012576/AnsiballZ_grafana_datasource.py && sleep 0'"'"''
<host1.net> (0, b'', b'')
<host1.net> ESTABLISH SSH CONNECTION FOR USER: None
<host1.net> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/ansible/.ansible/cp/9ec4a2bc8e -tt host1.net '/bin/sh -c '"'"'sudo -H -S  -p "[sudo via ansible, key=ikkampdiahqechocotwsyoktdaldlfcz] password:" -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-ikkampdiahqechocotwsyoktdaldlfcz ; /usr/bin/python3 /home/ansible/.ansible/tmp/ansible-tmp-1596772708.4799106-4131-56466493012576/AnsiballZ_grafana_datasource.py'"'"'"'"'"'"'"'"' && sleep 0'"'"''
Escalation succeeded
<host1.net> (0, b'\r\n\r\n{"failed": false, "msg": "Datasource authtest created : Datasource added", "changed": true, "name": "authtest", "id": 45, "invocation": {"module_args": {"user": "testuser", "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "name": "authtest", "url": "https://authtest:3306", "access": "proxy", "state": "present", "grafana_url": "https://host1.net", "validate_certs": false, "grafana_api_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "ds_type": "mysql", "database": "testdb", "basic_auth_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "basic_auth_user": "testuser", "use_proxy": true, "url_username": "admin", "url_password": "admin", "with_credentials": false, "tls_skip_verify": false, "is_default": false, "org_id": 1, "es_version": 5, "max_concurrent_shard_requests": 256, "time_field": "@timestamp", "interval": "", "tsdb_version": 1, "tsdb_resolution": "second", "sslmode": "disable", "trends": false, "aws_auth_type": "keys", "aws_default_region": "us-east-1", "aws_access_key": "", "aws_secret_key": "", "aws_credentials_profile": "", "aws_assume_role_arn": "", "client_cert": null, "client_key": null, "tls_client_cert": null, "tls_client_key": null, "tls_ca_cert": null, "time_interval": null, "aws_custom_metrics_namespaces": null}}}\r\n', b'Shared connection to host1.net closed.\r\n')
<host1.net> ESTABLISH SSH CONNECTION FOR USER: None
<host1.net> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/ansible/.ansible/cp/9ec4a2bc8e host1.net '/bin/sh -c '"'"'rm -f -r /home/ansible/.ansible/tmp/ansible-tmp-1596772708.4799106-4131-56466493012576/ > /dev/null 2>&1 && sleep 0'"'"''
<host1.net> (0, b'', b'')
host1.net | CHANGED => {
    "changed": true,
    "id": 45,
    "invocation": {
        "module_args": {
            "access": "proxy",
            "aws_access_key": "",
            "aws_assume_role_arn": "",
            "aws_auth_type": "keys",
            "aws_credentials_profile": "",
            "aws_custom_metrics_namespaces": null,
            "aws_default_region": "us-east-1",
            "aws_secret_key": "",
            "basic_auth_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "basic_auth_user": "testuser",
            "client_cert": null,
            "client_key": null,
            "database": "testdb",
            "ds_type": "mysql",
            "es_version": 5,
            "grafana_api_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "grafana_url": "https://host1.net",
            "interval": "",
            "is_default": false,
            "max_concurrent_shard_requests": 256,
            "name": "authtest",
            "org_id": 1,
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "sslmode": "disable",
            "state": "present",
            "time_field": "@timestamp",
            "time_interval": null,
            "tls_ca_cert": null,
            "tls_client_cert": null,
            "tls_client_key": null,
            "tls_skip_verify": false,
            "trends": false,
            "tsdb_resolution": "second",
            "tsdb_version": 1,
            "url": "https://authtest:3306",
            "url_password": "admin",
            "url_username": "admin",
            "use_proxy": true,
            "user": "testuser",
            "validate_certs": false,
            "with_credentials": false
        }
    },
    "msg": "Datasource authtest created : Datasource added",
    "name": "authtest"
}
META: ran handlers
META: ran handlers

```

**Comment by rockaut** [comment]
I think this is somewhat related to #107 and should be fixed then.

**Comment by rockaut** [comment]
Just to be clear: there is a problem around the secureJsonData (see #108, #107) yes ... but it also depends on the datasource!

For example the Postgres Datasource:
- if you set the `password: xxx` property it is unencrypted and the grafana gui doesn't show you it's set BUT it works anyway
- if you set a password in secureJsonData or use the secure_password (after my PR) the GUI show it correctly

This was changed some versions ago. I guess to force or bring the users to use encrypted data instead of plain text. But again it depends on the plugin/datasource itself - it has to support secureJsonData.password and if it doesn't it doesn't work. I don't know the Grafana SDK but I think I heard they now have a "fallback" for the plugins around this but it's out focus of this project tho cover that.

**Comment by rockaut** [comment]
#108  is merged now into master.

The following would provision the datasource with password as secureJsonData.

Be advised that (see module notes) that on subsequent runs the secure password isn't updated. For that you would additionally need `enforce_secure_data: true` additionally.

```
  - name: provisioning datasources
    grafana_datasource:
      name: "{{ item.name }}"
      database: "{{ item.database}}"
      url: dsurl:port
      grafana_url: "{{ hostvars[inventory_hostname].grafurl }}"
      validate_certs: no
      grafana_api_key: "{{ hostvars[inventory_hostname].grafapikey }}"
      ds_type: mysql
      # any of the following reproduces the issue:
      user: testuser
      additional_secure_json_data:
        password: test123
```

**Comment by rrey** [comment]
Hi @onefourfive,

I just released 1.0.0 which include a patch from @rockaut (thanks !) around password management.
Could you test this release and confirm that password is not updated as expected ?

Thanks in advance.

**Comment by onefourfive** [comment]
@rrey Sorry for the delay, I just had the opportunity to update & provision more datasources today.

This worked for me! Thank you for addressing this issue, it's a huge timesaver.
